cdc/sinkv2/ddlsink/mq/kafka_sink.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mq
+
+import (
+	"context"
+	"net/url"
+
+	"github.com/Shopify/sarama"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/cdc/sink/mq/dispatcher"
+	"github.com/pingcap/tiflow/cdc/sink/mq/producer/kafka"
+	"github.com/pingcap/tiflow/cdc/sinkv2/ddlsink/mq/producer"
+	mqutil "github.com/pingcap/tiflow/cdc/sinkv2/util/mq"
+	"github.com/pingcap/tiflow/pkg/config"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	pkafka "github.com/pingcap/tiflow/pkg/kafka"
+	"go.uber.org/zap"
+)
+
+// NewKafkaDDLSink will verify the config and create a Kafka DDL Sink.
+func NewKafkaDDLSink(
+	ctx context.Context,
+	sinkURI *url.URL,
+	replicaConfig *config.ReplicaConfig,
+	adminClientCreator pkafka.ClusterAdminClientCreator,
+	producerCreator producer.Factory,
+) (*ddlSink, error) {
+	topic, err := mqutil.GetTopic(sinkURI)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	baseConfig := kafka.NewConfig()
+	if err := baseConfig.Apply(sinkURI); err != nil {
+		return nil, cerror.WrapError(cerror.ErrKafkaInvalidConfig, err)
+	}
+	saramaConfig, err := kafka.NewSaramaConfig(ctx, baseConfig)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	adminClient, err := adminClientCreator(baseConfig.BrokerEndpoints, saramaConfig)
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrKafkaNewSaramaProducer, err)
+	}
+	// we must close adminClient when this func return cause by an error
+	// otherwise the adminClient will never be closed and lead to a goroutine leak
+	// TODO: we should pass this adminClient to the producer to get the metrics.
+	defer func() {
+		_ = adminClient.Close()
+	}()
```

Comment:
Should we still close it if the function succeeds?

Reply:
For now, yes. I will deal with it in a later PR.